include/ginkgo/core/base/segmented_array.hpp

```diff
@@ -0,0 +1,236 @@
+     * @param sizes  the sizes of each segment
+     */
+    segmented_array(std::shared_ptr<const Executor> exec,
+                    const std::vector<int64>& sizes);
+
+    /**
+     * Creates a segmented array from a flat buffer and segment sizes.
+     *
+     * @param buffer  the flat buffer, it's size has to match the sum of sizes
+     * @param sizes  the sizes of each segment
+     */
+    segmented_array(gko::array<T> buffer, const std::vector<int64>& sizes);
+
+    /**
+     * Creates a segmented array from a flat buffer and offsets.
+     *
+     * @param buffer  the flat buffer, it's size has to match the last element
+     *                of offsets
+     * @param offsets  the index offsets for each segment, and the total size of
+     *                 the buffer as last element
+     */
+    segmented_array(gko::array<T> buffer, gko::array<int64> offsets);
+
+    /**
+     * Creates an uninitialized segmented array from offsets.
+     *
+     * @param offsets  the index offsets for each segment, and the total size of
+     *                 the buffer as last element
+     */
+    explicit segmented_array(gko::array<int64> offsets);
+
+    /**
+     * Copies a segmented array to a different executor.
+     *
+     * @param exec  the executor to copy to
+     * @param other  the segmented array to copy from
+     */
+    segmented_array(std::shared_ptr<const Executor> exec,
+                    const segmented_array& other);
+
+    /**
+     * Moves a segmented array to a different executor.
+     *
+     * @param exec  the executor to move to
+     * @param other  the segmented array to move from
+     */
+    segmented_array(std::shared_ptr<const Executor> exec,
+                    segmented_array&& other);
+
+    /**
+     * Get the total size of the stored buffer.
+     *
+     * @return  the total size of the stored buffer.
+     */
+    size_type get_size() const;
+
+    /**
+     * Get the number of segments.
+     *
+     * @return  the number of segments
+     */
+    size_type get_segment_size() const;
+
+    /**
+     * Access to the flat buffer.
+     *
+     * @return  the flat buffer
+     */
+    array<T>& get_flat();
+
+    /**
+     * Const-access to the flat buffer
+     *
+     * @return  the flat buffer
+     */
+    const array<T>& get_flat() const;
+
+    /**
+     * Access to the segment offsets.
+     *
+     * @return  the segment offsets
+     */
+    const gko::array<int64>& get_offsets() const;
+
+    /**
+     * Access the executor.
+     *
+     * @return  the executor
+     */
+    std::shared_ptr<const Executor> get_executor() const;
+
+private:
+    gko::array<T> buffer_;
+    gko::array<int64> offsets_;
+};
+
+
+template <typename T>
+size_type segmented_array<T>::get_size() const
+{
+    return buffer_.get_size();
+}
+
+
+template <typename T>
+size_type segmented_array<T>::get_segment_size() const
+{
+    return offsets_.get_size() ? offsets_.get_size() - 1 : 0;
```

Comment:
same here, also should that be segment_count?

Reply:
I'm getting a bit tired of all the rebasing, would it be ok to leave the name as it is?

Reply:
changed it, since there were other updates anyway.